win: port blk_rw unittest

ChandKV opened the pull request:
- Fix mmap to reserve map till end of file and reserve the virtual
  address for the remaining length, when given length is greater than
  file length.
   o Add dependency to ntdll.lib to call the native (or NT) function
   o Update win_mmap tests
- Fix util_replica_create_local and util_replica_open_local to retry
  its attemp to map multiple parts into a contiguous virtual address
  space, say 10 times.  This is required because our implementation of
  mmap() might fail to map the file in the requested address even if
  it's already reserved.
- Fix runtests.ps1 to define usage above the functions that call it
- Implement convert_to_bytes helper and fix create_poolset in Windows
  to honor (K/M/G) suffixes, like Linux
- Convert the redundant check in util_map_part that verifies tha mmap
  correctly fails if it can't map in the given address if MAP_FIXED is
  set
- Fix create_poolset's drive letter detection logic
- Fix the projects configurations to:
   o have build the same configuration that's chosen
   o make sure that all configurations are buildable
- Fix Copyright line for Microsoft contributions

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/pmem/nvml/1031)
<!-- Reviewable:end -->

krzycz commented:
Reviewed 97 of 100 files at r1.
Review status: 97 of 100 files reviewed at latest revision, 10 unresolved discussions, some commit checks failed.

---

*[src/test/blk_rw/err0w.log.match, line 0 \[r1\]](https://reviewable.io:443/reviews/pmem/nvml/1031#-KNaWV-8N4B56jU2HdaU:-KNaWV-8N4B56jU2HdaV:643728168) ([raw file](https://github.com/pmem/nvml/blob/a1b4e45a97ba5f200302af19bf137d53cda5b88d/src/test/blk_rw/err0w.log.match#L0)):*
Are all of those needed? For me, it looks like only err8w.log and err9w.log are necessary.

---

*[src/test/blk_rw/out5.log.match, line 24 \[r1\]](https://reviewable.io:443/reviews/pmem/nvml/1031#-KNaQPu7cwpoO6nEn6N2:-KNaQPu7cwpoO6nEn6N3:555104500) ([raw file](https://github.com/pmem/nvml/blob/6e53c028cd41db4065308f1bd00178402847e68b/src/test/blk_rw/out5.log.match#L24)):*
> ```
> set_error lba 300
> write     lba 400: {8}
> read      lba 100: Input$(nW)output error
> ```

Is there really backslash (or something else) in this error message? Or it's just a result of just find&replace...

---

*[src/test/blk_rw/TEST0.PS1, line 67 \[r1\]](https://reviewable.io:443/reviews/pmem/nvml/1031#-KNaRMidHmEzM0lAdAa1:-KNaRMidHmEzM0lAdAa2:1888571527) ([raw file](https://github.com/pmem/nvml/blob/6e53c028cd41db4065308f1bd00178402847e68b/src/test/blk_rw/TEST0.PS1#L67)):*
> ```
> # Attempts to zero uninitialized blocks are nops (should succeed).
> #
> expect_normal_exit ..\..\x64\debug\blk_rw$EXESUFFIX 512 $DIR\testfile1 c `
> ```

Please, use $Env:EXE_DIR instead of hardcoded path (same in other TEST scripts).

---

*[src/test/blk_rw_mt/blk_rw_mt.c, line 3 \[r1\]](https://reviewable.io:443/reviews/pmem/nvml/1031#-KNaY4XwkDvncdyPXC7b:-KNaY4XwkDvncdyPXC7c:-205174991) ([raw file](https://github.com/pmem/nvml/blob/6e53c028cd41db4065308f1bd00178402847e68b/src/test/blk_rw_mt/blk_rw_mt.c#L3)):*
> ```C
> /*
>  * Copyright 2014-2016, Intel Corporation
>  * Copyright (c) Microsoft Corporation. All rights reserved.
> ```

??? ;-)

---

*[src/test/blk_rw_mt/blk_rw_mt.vcxproj, line 103 \[r1\]](https://reviewable.io:443/reviews/pmem/nvml/1031#-KNaS877GAN95jMIoVjt:-KNaS877GAN95jMIoVju:1323741427) ([raw file](https://github.com/pmem/nvml/blob/6e53c028cd41db4065308f1bd00178402847e68b/src/test/blk_rw_mt/blk_rw_mt.vcxproj#L103)):*
> ```XML
>       <SubSystem>Console</SubSystem>
>       <GenerateDebugInformation>true</GenerateDebugInformation>
>       <AdditionalDependencies>Shlwapi.lib;kernel32.lib;user32.lib;gdi32.lib;winspool.lib;comdlg32.lib;advapi32.lib;shell32.lib;ole32.lib;oleaut32.lib;uuid.lib;odbc32.lib;odbccp32.lib;%(AdditionalDependencies)</AdditionalDependencies>
> ```

Are all of those libs really needed here?

---

*[src/test/tools/pmemalloc/pmemalloc.vcxproj, line 161 \[r1\]](https://reviewable.io:443/reviews/pmem/nvml/1031#-KNaXcYaH0Q8fxwMtuwC:-KNaXcYaH0Q8fxwMtuwD:1347035505) ([raw file](https://github.com/pmem/nvml/blob/6e53c028cd41db4065308f1bd00178402847e68b/src/test/tools/pmemalloc/pmemalloc.vcxproj#L161)):*
> ```XML
>     <Link>
>       <SubSystem>Console</SubSystem>
>       <AdditionalLibraryDirectories>$(OutDir);$(OutDir)/../Static-Release;C:\Users\Administrator\nvml\src\windows\getopt_port\x64\Debug;$(ProjectDir)..\..\..\windows\getops_port\x64\Debug;$(ProjectDir)..\..\..\windows\getops_port\x64\Debug\</AdditionalLibraryDirectories>
> ```

hardcoded and incorrect paths

---

*[src/test/tools/pmemalloc/pmemalloc.vcxproj, line 183 \[r1\]](https://reviewable.io:443/reviews/pmem/nvml/1031#-KNaXUCo258Ei6jHIhTY:-KNaXUCo258Ei6jHIhTZ:1347035505) ([raw file](https://github.com/pmem/nvml/blob/6e53c028cd41db4065308f1bd00178402847e68b/src/test/tools/pmemalloc/pmemalloc.vcxproj#L183)):*
> ```XML
>     <Link>
>       <SubSystem>Console</SubSystem>
>       <AdditionalLibraryDirectories>$(OutDir);$(OutDir)/../Static-Release;C:\Users\Administrator\nvml\src\windows\getopt_port\x64\Debug;$(ProjectDir)..\..\..\windows\getops_port\x64\Debug;$(ProjectDir)..\..\..\windows\getops_port\x64\Debug\</AdditionalLibraryDirectories>
> ```

hardcoded and incorrect paths

---

*[src/test/tools/pmemspoil/pmemspoil.vcxproj, line 38 \[r1\]](https://reviewable.io:443/reviews/pmem/nvml/1031#-KNaTOfUmhwtFG95-ecZ:-KNaTOfUmhwtFG95-ec_:-591181398) ([raw file](https://github.com/pmem/nvml/blob/6e53c028cd41db4065308f1bd00178402847e68b/src/test/tools/pmemspoil/pmemspoil.vcxproj#L38)):*
> ```XML
>     </ProjectReference>
>     <ProjectReference Include="..\..\..\libpmempool\libpmempool.vcxproj">
>       <Project>{cf9a0883-6334-44c7-ac29-349468c78e27}</Project>
> ```

pmemspoil should not depend on pmempool/libpmempool

---

*[src/test/tools/pmemspoil/pmemspoil.vcxproj, line 195 \[r1\]](https://reviewable.io:443/reviews/pmem/nvml/1031#-KNaTZ5u8Q0eePFSg5I3:-KNaTZ5u8Q0eePFSg5I4:517301890) ([raw file](https://github.com/pmem/nvml/blob/6e53c028cd41db4065308f1bd00178402847e68b/src/test/tools/pmemspoil/pmemspoil.vcxproj#L195)):*
> ```XML
>     <Link>
>       <SubSystem>Console</SubSystem>
>       <AdditionalLibraryDirectories>$(OutDir);$(OutDir)/../Static-Release;C:\Users\Administrator\nvml\src\windows\getopt_port\x64\Debug;$(ProjectDir)..\..\..\windows\getops_port\x64\Debug;$(ProjectDir)..\..\..\windows\getops_port\x64\Debug\</AdditionalLibraryDirectories>
> ```

hardcoded (and wrong) paths

---

*[src/test/unittest/unittest.ps1, line 63 \[r1\]](https://reviewable.io:443/reviews/pmem/nvml/1031#-KNaUO3omsModFwn6ffF:-KNaUO3omsModFwn6ffG:-1238311594) ([raw file](https://github.com/pmem/nvml/blob/6e53c028cd41db4065308f1bd00178402847e68b/src/test/unittest/unittest.ps1#L63)):*
> ```PowerShell
> 
> #
> # convert_to_bytes -- converts the sting with K, M, G or T prefixes
> ```

string

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/nvml/1031)*
<!-- Sent from Reviewable.io -->

lplewa commented:
Reviewed 2 of 100 files at r1.
Review status: 98 of 100 files reviewed at latest revision, 16 unresolved discussions, some commit checks failed.

---

*[src/NVML.sln, line 3 \[r1\]](https://reviewable.io:443/reviews/pmem/nvml/1031#-KNaYYKgbgU7Amvzjgty:-KNaYYKhDcaPxi4y1h8r:-1176189968) ([raw file](https://github.com/pmem/nvml/blob/6e53c028cd41db4065308f1bd00178402847e68b/src/NVML.sln#L3)):*
> ```
> Microsoft Visual Studio Solution File, Format Version 12.00
> # Visual Studio 14
> VisualStudioVersion = 14.0.25123.0
> ```

Please do not update this line

---

*[src/common/set.c, line 1668 \[r1\]](https://reviewable.io:443/reviews/pmem/nvml/1031#-KNaaZZezzcRvKg9be4g:-KNaaZZezzcRvKg9be4h:1026023266) ([raw file](https://github.com/pmem/nvml/blob/6e53c028cd41db4065308f1bd00178402847e68b/src/common/set.c#L1668)):*
> ```C
> 					/* unmap the parts that are mapped */
> 					for (unsigned p = 0; p < p; p++)
> 						util_unmap_part(&rep->part[p]);
> ```

Try consider to split this function (or refactor) as you have here 6 levels of indentation

---

*[src/test/unittest/unittest.ps1, line 63 \[r1\]](https://reviewable.io:443/reviews/pmem/nvml/1031#-KNap-c-eRR5BnUgj7yn:-KNap-c0u-XwYv-kGe7Q:294197220) ([raw file](https://github.com/pmem/nvml/blob/6e53c028cd41db4065308f1bd00178402847e68b/src/test/unittest/unittest.ps1#L63)):*
> ```PowerShell
> 
> #
> # convert_to_bytes -- converts the sting with K, M, G or T prefixes
> ```

Suffixes

---

*[src/test/unittest/unittest.ps1, line 74 \[r1\]](https://reviewable.io:443/reviews/pmem/nvml/1031#-KNay4D059ltyRJOtsjS:-KNay4D059ltyRJOtsjT:1838348226) ([raw file](https://github.com/pmem/nvml/blob/6e53c028cd41db4065308f1bd00178402847e68b/src/test/unittest/unittest.ps1#L74)):*
> ```PowerShell
> #
> 
> function convert_to_bytes() {
> ```

Please consider to support all suffixes from util_parse_size function

---

*[src/test/unittest/unittest.ps1, line 289 \[r1\]](https://reviewable.io:443/reviews/pmem/nvml/1031#-KNanbc0bg6LAcpVTF32:-KNanbc0bg6LAcpVTF33:-1508819271) ([raw file](https://github.com/pmem/nvml/blob/6e53c028cd41db4065308f1bd00178402847e68b/src/test/unittest/unittest.ps1#L289)):*
> ```PowerShell
>             # non-zeroed file
>             'n' { create_file $asize $fpath }
>             # non-zeroed file, except 4K header
> ```

These functions still assumes size in MB+KB style

when changing it please check Linux version of these functions

---

*[src/test/unittest/unittest.ps1, line 297 \[r1\]](https://reviewable.io:443/reviews/pmem/nvml/1031#-KNaoUN3M9nnpdlJSG-y:-KNaoUN3M9nnpdlJSG-z:-332744155) ([raw file](https://github.com/pmem/nvml/blob/6e53c028cd41db4065308f1bd00178402847e68b/src/test/unittest/unittest.ps1#L297)):*
> ```PowerShell
>         #	fi
> 
>         echo "$fsize $fpath" | out-file -Append -encoding ASCII $psfile
> ```

NVML poolsets supports size with suffix so it will be better to put non-converted version in to poolsetfile

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/nvml/1031)*
<!-- Sent from Reviewable.io -->

lplewa commented:
Refs: pmem/issues#214

ChandKV commented:
Review status: 77 of 101 files reviewed at latest revision, 16 unresolved discussions.

---

*[src/NVML.sln, line 3 \[r1\]](https://reviewable.io:443/reviews/pmem/nvml/1031#-KNaYYKgbgU7Amvzjgty:-KNbcoBqBEVCKzUumZWm:822937724) ([raw file](https://github.com/pmem/nvml/blob/6e53c028cd41db4065308f1bd00178402847e68b/src/NVML.sln#L3)):*
<details><summary><i>Previously, lplewa (Łukasz Plewa) wrote…</i></summary>
> Please do not update this line
</details>
Sure let me remove this. Just so you know, I removed that change while working on it, missed it on the final change. Thank you.

---

*[src/common/set.c, line 1668 \[r1\]](https://reviewable.io:443/reviews/pmem/nvml/1031#-KNaaZZezzcRvKg9be4g:-KNc1HW56PhjiFTHFP8H:-1751292471) ([raw file](https://github.com/pmem/nvml/blob/6e53c028cd41db4065308f1bd00178402847e68b/src/common/set.c#L1668)):*
<details><summary><i>Previously, lplewa (Łukasz Plewa) wrote…</i></summary>
> Try consider to split this function (or refactor) as you have here 6 levels of indentation
</details>
Moved the sixth level to a helper function along with fixing a bad bug in it.

Trying to reduce further levels looked rather odd, let me know if we should reduce further.

---

*[src/test/blk_rw/err0w.log.match, line  \[r1\]](https://reviewable.io:443/reviews/pmem/nvml/1031#-KNaWV-8N4B56jU2HdaU:-KNbdV-7DnOUUPS-qQ2m:-1933205192) ([raw file](https://github.com/pmem/nvml/blob/a1b4e45a97ba5f200302af19bf137d53cda5b88d/src/test/blk_rw/err0w.log.match#L)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary>
> Are all of those needed? For me, it looks like only err8w.log and err9w.log are necessary.
</details>
Linux had err.log.match, and I was thinking that it's required to make sure that the test doesn't generate any ERR logs. Isn't that true?

---

*[src/test/blk_rw/out5.log.match, line 24 \[r1\]](https://reviewable.io:443/reviews/pmem/nvml/1031#-KNaQPu7cwpoO6nEn6N2:-KNbds0bxycNT6Q_n6MS:-930869581) ([raw file](https://github.com/pmem/nvml/blob/6e53c028cd41db4065308f1bd00178402847e68b/src/test/blk_rw/out5.log.match#L24)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary>
> Is there really backslash (or something else) in this error message? Or it's just a result of just find&replace...
</details>
Thank you, find and replace change :), let me revert these.

---

*[src/test/blk_rw/TEST0.PS1, line 67 \[r1\]](https://reviewable.io:443/reviews/pmem/nvml/1031#-KNaRMidHmEzM0lAdAa1:-KNbeLls9d3T5i81pu10:1374766494) ([raw file](https://github.com/pmem/nvml/blob/6e53c028cd41db4065308f1bd00178402847e68b/src/test/blk_rw/TEST0.PS1#L67)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary>
> Please, use $Env:EXE_DIR instead of hardcoded path (same in other TEST scripts).
</details>
Ah... let me switch to that, thank you.

---

*[src/test/blk_rw_mt/blk_rw_mt.c, line 3 \[r1\]](https://reviewable.io:443/reviews/pmem/nvml/1031#-KNaY4XwkDvncdyPXC7b:-KNbeTGCsUCuqCMbfvKr:891760394) ([raw file](https://github.com/pmem/nvml/blob/6e53c028cd41db4065308f1bd00178402847e68b/src/test/blk_rw_mt/blk_rw_mt.c#L3)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary>
> ??? ;-)
</details>
I didn't add the same with my last checkin. So included the same now. I need some practice to do that may be :).

---

*[src/test/blk_rw_mt/blk_rw_mt.vcxproj, line 103 \[r1\]](https://reviewable.io:443/reviews/pmem/nvml/1031#-KNaS877GAN95jMIoVjt:-KNbevKUG7BwYeyEBm8Y:1675259509) ([raw file](https://github.com/pmem/nvml/blob/6e53c028cd41db4065308f1bd00178402847e68b/src/test/blk_rw_mt/blk_rw_mt.vcxproj#L103)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary>
> Are all of those libs really needed here?
</details>
Seems like <inherit from the parent or project defaults> is bring those hidden treasure. May be there are couple of other projects to cleanup. Fixed this project for now.

---

*[src/test/tools/pmemalloc/pmemalloc.vcxproj, line 161 \[r1\]](https://reviewable.io:443/reviews/pmem/nvml/1031#-KNaXcYaH0Q8fxwMtuwC:-KNbguwML3pd4Eyx3ijs:-234154027) ([raw file](https://github.com/pmem/nvml/blob/6e53c028cd41db4065308f1bd00178402847e68b/src/test/tools/pmemalloc/pmemalloc.vcxproj#L161)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary>
> hardcoded and incorrect paths
</details>
I copied this from other configuration of the same project. Let me remove the hard coded ones, though.

But these projects need more fixup. Currently, couple of these projects doesn't build for Static-Debug and Static-Release (because they redefine a function available in another library but it's not exported) and I have excluded them from build for these configurations.

---

*[src/test/tools/pmemalloc/pmemalloc.vcxproj, line 183 \[r1\]](https://reviewable.io:443/reviews/pmem/nvml/1031#-KNaXUCo258Ei6jHIhTY:-KNbukADtJ6rDlfXa4ns:-1981775876) ([raw file](https://github.com/pmem/nvml/blob/6e53c028cd41db4065308f1bd00178402847e68b/src/test/tools/pmemalloc/pmemalloc.vcxproj#L183)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary>
> hardcoded and incorrect paths
</details>
Same as the comment for pmemalloc. Fixed.

---

*[src/test/tools/pmemspoil/pmemspoil.vcxproj, line 38 \[r1\]](https://reviewable.io:443/reviews/pmem/nvml/1031#-KNaTOfUmhwtFG95-ecZ:-KNcBieeus8zng5lrC6m:998414963) ([raw file](https://github.com/pmem/nvml/blob/6e53c028cd41db4065308f1bd00178402847e68b/src/test/tools/pmemspoil/pmemspoil.vcxproj#L38)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary>
> pmemspoil should not depend on pmempool/libpmempool
</details>
Removed the same, thank you.

---

*[src/test/tools/pmemspoil/pmemspoil.vcxproj, line 195 \[r1\]](https://reviewable.io:443/reviews/pmem/nvml/1031#-KNaTZ5u8Q0eePFSg5I3:-KNbuxDSxC2tI4aQv29d:-1981775876) ([raw file](https://github.com/pmem/nvml/blob/6e53c028cd41db4065308f1bd00178402847e68b/src/test/tools/pmemspoil/pmemspoil.vcxproj#L195)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary>
> hardcoded (and wrong) paths
</details>
Same as the comment for pmemalloc. Fixed.

---

*[src/test/unittest/unittest.ps1, line 63 \[r1\]](https://reviewable.io:443/reviews/pmem/nvml/1031#-KNaUO3omsModFwn6ffF:-KNbv8rR_fnvanmXjRPQ:-412870703) ([raw file](https://github.com/pmem/nvml/blob/6e53c028cd41db4065308f1bd00178402847e68b/src/test/unittest/unittest.ps1#L63)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary>
> string
</details>
Thank you, let me fix that.

---

*[src/test/unittest/unittest.ps1, line 63 \[r1\]](https://reviewable.io:443/reviews/pmem/nvml/1031#-KNap-c-eRR5BnUgj7yn:-KNbvD6rJ36K5-paX6wE:-946459406) ([raw file](https://github.com/pmem/nvml/blob/6e53c028cd41db4065308f1bd00178402847e68b/src/test/unittest/unittest.ps1#L63)):*
<details><summary><i>Previously, lplewa (Łukasz Plewa) wrote…</i></summary>
> Suffixes
</details>
My bad :), thank you.

---

*[src/test/unittest/unittest.ps1, line 74 \[r1\]](https://reviewable.io:443/reviews/pmem/nvml/1031#-KNay4D059ltyRJOtsjS:-KNcDyxhHM-Rcua9zghC:1525508201) ([raw file](https://github.com/pmem/nvml/blob/6e53c028cd41db4065308f1bd00178402847e68b/src/test/unittest/unittest.ps1#L74)):*
<details><summary><i>Previously, lplewa (Łukasz Plewa) wrote…</i></summary>
> Please consider to support all suffixes from util_parse_size function
</details>
Implementing suffixes "kB, MB, GB, TB" would be interesting as powershell generally converts them to "byte" values using 1024 as the multiplier. And util_parse_size() wants to use "1000" as the multiplier. convert_to_bytes() expects a string so 1kB will remain as is and will be converted to 1024, but the callers / users should make sure that the value is not expanded before reaching this function.

Btw, in "kB" is the lower case "k" intended in util_parse_size(). Also, should convert_to_bytes() be case sensitive as well?

---

*[src/test/unittest/unittest.ps1, line 289 \[r1\]](https://reviewable.io:443/reviews/pmem/nvml/1031#-KNanbc0bg6LAcpVTF32:-KNbvcMh9uSLuL4RZM22:2143672576) ([raw file](https://github.com/pmem/nvml/blob/6e53c028cd41db4065308f1bd00178402847e68b/src/test/unittest/unittest.ps1#L289)):*
<details><summary><i>Previously, lplewa (Łukasz Plewa) wrote…</i></summary>
> These functions still assumes size in MB+KB style
> 
> when changing it please check Linux version of these functions
</details>
Sure, that will be one of my follow up changes.

---

*[src/test/unittest/unittest.ps1, line 297 \[r1\]](https://reviewable.io:443/reviews/pmem/nvml/1031#-KNaoUN3M9nnpdlJSG-y:-KNbwlwpM5-VDfydqypW:405350936) ([raw file](https://github.com/pmem/nvml/blob/6e53c028cd41db4065308f1bd00178402847e68b/src/test/unittest/unittest.ps1#L297)):*
<details><summary><i>Previously, lplewa (Łukasz Plewa) wrote…</i></summary>
> NVML poolsets supports size with suffix so it will be better to put non-converted version in to poolsetfile
</details>
I see. Sure, let me do that. Thank you. Fixed.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/nvml/1031)*
<!-- Sent from Reviewable.io -->

lplewa commented:
Reviewed 1 of 100 files at r1.
Review status: 78 of 101 files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[src/test/unittest/unittest.ps1, line 74 \[r1\]](https://reviewable.io:443/reviews/pmem/nvml/1031#-KNay4D059ltyRJOtsjS:-KNfUx7YKxchOlal7lyK:-294798276) ([raw file](https://github.com/pmem/nvml/blob/6e53c028cd41db4065308f1bd00178402847e68b/src/test/unittest/unittest.ps1#L74)):*
<details><summary><i>Previously, ChandKV (Chandra Kumar) wrote…</i></summary>
> Implementing suffixes "kB, MB, GB, TB" would be interesting as powershell generally converts them to "byte" values using 1024 as the multiplier. And util_parse_size() wants to use "1000" as the multiplier. convert_to_bytes() expects a string so 1kB will remain as is and will be converted to 1024, but the callers / users should make sure that the value is not expanded before reaching this function.
> 
> Btw, in "kB" is the lower case "k" intended in util_parse_size(). Also, should convert_to_bytes() be case sensitive as well?
</details>
Yes it is intended. SI suffixes are lower case up to kilo (k) and upper case above kilo.
https://en.wikipedia.org/wiki/Metric_prefix#List_of_SI_prefixes

Lets left this discussion open as i'm not sure if we should support this same unit convention on Windows and Linux.

From one side it is nice to be cross platform compatible, but from second hand Windows users are not familiar with suffixes commonly used in Linux (and in our library)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/nvml/1031)*
<!-- Sent from Reviewable.io -->

krzycz commented:
A general question - is ntdll.lib required by all the projects or just in libpmem (which exports mmap/mnumap APIs)?

---

Reviewed 2 of 100 files at r1, 21 of 22 files at r2, 1 of 1 files at r3.
Review status: all files reviewed at latest revision, 18 unresolved discussions, some commit checks failed.

---

*[src/common/set.c, line 1655 \[r1\]](https://reviewable.io:443/reviews/pmem/nvml/1031#-KNab9INnJG1FeL_HYWu:-KNab9INnJG1FeL_HYWv:-286088937) ([raw file](https://github.com/pmem/nvml/blob/6e53c028cd41db4065308f1bd00178402847e68b/src/common/set.c#L1655)):*
> ```C
> 					flags | MAP_FIXED) != 0) {
> 				/*
> 				 * if we can't map the part at the address we
> ```

If the first call to util_map_part() reserves space for all the parts (with VirtualAlloc), then why could subsequent mappings fail? Does it ever happen in practice?

---

*[src/common/set.c, line 1667 \[r1\]](https://reviewable.io:443/reviews/pmem/nvml/1031#-KNa_DQCKpN5G5zn9vfu:-KNa_DQCKpN5G5zn9vfv:146362660) ([raw file](https://github.com/pmem/nvml/blob/6e53c028cd41db4065308f1bd00178402847e68b/src/common/set.c#L1667)):*
> ```C
> 
> 					/* unmap the parts that are mapped */
> 					for (unsigned p = 0; p < p; p++)
> ```

`p < p` ???
I think you should use some other var name for a loop counter.

---

*[src/common/set.c, line 1673 \[r1\]](https://reviewable.io:443/reviews/pmem/nvml/1031#-KNa_zLZ2MDWqkaobUUn:-KNa_zLZ2MDWqkaobUUo:-300557148) ([raw file](https://github.com/pmem/nvml/blob/6e53c028cd41db4065308f1bd00178402847e68b/src/common/set.c#L1673)):*
> ```C
> 					ASSERTne(addr, NULL);
> 					ASSERTne(addr, MAP_FAILED);
> 					munmap(addr, 0);
> ```

This is confusing. Shouldn't it be: `munmap(addr, rep->repsize);` 

---

*[src/common/set.c, line 1705 \[r1\]](https://reviewable.io:443/reviews/pmem/nvml/1031#-KNaaEZ33XWVcvzb6DHn:-KNaaEZ33XWVcvzb6DHo:-565176780) ([raw file](https://github.com/pmem/nvml/blob/6e53c028cd41db4065308f1bd00178402847e68b/src/common/set.c#L1705)):*
> ```C
> 		ASSERTne(addr, NULL);
> 		ASSERTne(addr, MAP_FAILED);
> 		munmap(addr, 0);
> ```

as above

---

*[src/common/set.c, line 1605 \[r3\]](https://reviewable.io:443/reviews/pmem/nvml/1031#-KNfgC9ZbwNSu9egJ2dZ:-KNfgC9ZbwNSu9egJ2d_:-70606287) ([raw file](https://github.com/pmem/nvml/blob/589413cb7c39f0161de16ed05f05d79ad746f24c/src/common/set.c#L1605)):*
> ```C
> 		prev_repl_uuid, next_repl_uuid, arch_flags);
> 
> 	int remaining_retries = 10;
> ```

I'm not sure if retries make any sense on Linux, as it would either succeed on the first attempt (likely) or fail (if there is really no unused region in process address space). In other words, for Linux this counter could be 1, I believe.

On the other hand, I think we should refactor this code anyway, and reserve the space for _all_ the replicas first, then start to mapping them to memory. It's very unlikely we'll run out of space in case of 64-bit (or 48-bit) addressing, but...
imagine we have only 3TB contiguous unused region and two replicas - each of 1TB in size. If the first one is placed in a middle of the unused region, the second one could not be mapped...

Perhaps this is a subject for another patch - just leave an XXX comment on that to not forget about that.

---

*[src/test/blk_rw/err0w.log.match, line  \[r1\]](https://reviewable.io:443/reviews/pmem/nvml/1031#-KNaWV-8N4B56jU2HdaU:-KNfTzBfeINyIm547AOf:-1043817253) ([raw file](https://github.com/pmem/nvml/blob/a1b4e45a97ba5f200302af19bf137d53cda5b88d/src/test/blk_rw/err0w.log.match#L)):*
<details><summary><i>Previously, ChandKV (Chandra Kumar) wrote…</i></summary>
> Linux had err.log.match, and I was thinking that it's required to make sure that the test doesn't generate any ERR logs. Isn't that true?
</details>
I mean, for tests 1-7 output is the same on Linux and Windows, so we don't need to have two match files for each of those (with and without "w" suffix). This is required only for tests 8-9. 

---

*[src/test/unittest/unittest.ps1, line 74 \[r1\]](https://reviewable.io:443/reviews/pmem/nvml/1031#-KNay4D059ltyRJOtsjS:-KNfXJYyOpi7mNH6i4M7:380002463) ([raw file](https://github.com/pmem/nvml/blob/6e53c028cd41db4065308f1bd00178402847e68b/src/test/unittest/unittest.ps1#L74)):*
<details><summary><i>Previously, ChandKV (Chandra Kumar) wrote…</i></summary>
> Implementing suffixes "kB, MB, GB, TB" would be interesting as powershell generally converts them to "byte" values using 1024 as the multiplier. And util_parse_size() wants to use "1000" as the multiplier. convert_to_bytes() expects a string so 1kB will remain as is and will be converted to 1024, but the callers / users should make sure that the value is not expanded before reaching this function.
> 
> Btw, in "kB" is the lower case "k" intended in util_parse_size(). Also, should convert_to_bytes() be case sensitive as well?
</details>
Well, remember that util_parse_size() is a different creature - it accepts size units as described in our man pages. This function however, may pass the size (unmodified) to `truncate` command (on Linux), which has slightly different semantics than util_parse_size().
Anyway, the new version looks good, and it seems to be more or less compatible with `truncate`.

---

*[src/test/win_mmap/win_mmap.c, line 265 \[r3\]](https://reviewable.io:443/reviews/pmem/nvml/1031#-KNfeY23lD12gM39q6mJ:-KNfeY23lD12gM39q6mK:-2066446587) ([raw file](https://github.com/pmem/nvml/blob/589413cb7c39f0161de16ed05f05d79ad746f24c/src/test/win_mmap/win_mmap.c#L265)):*
> ```C
> #ifndef _WIN32
> 	UT_ASSERTne(ptr, MAP_FAILED);
> 	munmap(ptr, 0);
> ```

Why `0`? And why not `check_mapping()` ?

---

*[src/windows/win_mmap.c, line 153 \[r3\]](https://reviewable.io:443/reviews/pmem/nvml/1031#-KNfu9ycOqGx72IBB65B:-KNfu9ydDNmrsJYrP1VF:-961814629) ([raw file](https://github.com/pmem/nvml/blob/589413cb7c39f0161de16ed05f05d79ad746f24c/src/windows/win_mmap.c#L153)):*
> ```C
> 
> /*
>  * mfree_reservation -- frees the range that's previously reserved
> ```

... -- (internal) ...

---

*[src/windows/win_mmap.c, line 155 \[r3\]](https://reviewable.io:443/reviews/pmem/nvml/1031#-KNfu8D10MEwiUb3kBM0:-KNfu8D10MEwiUb3kBM1:-1535746558) ([raw file](https://github.com/pmem/nvml/blob/589413cb7c39f0161de16ed05f05d79ad746f24c/src/windows/win_mmap.c#L155)):*
> ```C
>  * mfree_reservation -- frees the range that's previously reserved
>  */
> int
> ```

static int

---

*[src/windows/win_mmap.c, line 284 \[r3\]](https://reviewable.io:443/reviews/pmem/nvml/1031#-KNfkkFP-Ect-6FT8e7Q:-KNfkkFP-Ect-6FT8e7R:-1115820044) ([raw file](https://github.com/pmem/nvml/blob/589413cb7c39f0161de16ed05f05d79ad746f24c/src/windows/win_mmap.c#L284)):*
> ```C
> 		/*
> 		 * in our implementation we are choosing to ignore fd when
> 		 * MAN_ANON is set, instead of failing.
> ```

MAP_ANON

---

*[src/windows/win_mmap.c, line 344 \[r3\]](https://reviewable.io:443/reviews/pmem/nvml/1031#-KNfphNK7DhiYlgbnN5l:-KNfphNK7DhiYlgbnN5m:-420193738) ([raw file](https://github.com/pmem/nvml/blob/589413cb7c39f0161de16ed05f05d79ad746f24c/src/windows/win_mmap.c#L344)):*
> ```C
> 			}
> 
> 			addr = reserved_addr;
> ```

Not needed, I think, as addr == reserved_addr at this point.

---

*[src/windows/win_mmap.c, line 401 \[r3\]](https://reviewable.io:443/reviews/pmem/nvml/1031#-KNfnhMSQvZ0GygzAMpR:-KNfnhMSQvZ0GygzAMpS:158852600) ([raw file](https://github.com/pmem/nvml/blob/589413cb7c39f0161de16ed05f05d79ad746f24c/src/windows/win_mmap.c#L401)):*
> ```C
>      * the call if MAP_FIXED is set
>      */
>     if (addr != NULL && base != addr && (flags & MAP_FIXED)) {
> ```

Isn't an assert enough? Seems like it's already handled above (except for checking if `base == addr`).

---

*[src/windows/win_mmap.c, line 444 \[r3\]](https://reviewable.io:443/reviews/pmem/nvml/1031#-KNfsMVCvx2bfoCqPTEB:-KNfsMVCvx2bfoCqPTEC:-1073541604) ([raw file](https://github.com/pmem/nvml/blob/589413cb7c39f0161de16ed05f05d79ad746f24c/src/windows/win_mmap.c#L444)):*
> ```C
> 	PFILE_MAPPING_TRACKER mtb = NULL;
> 	PFILE_MAPPING_TRACKER mte = NULL;
> 	HANDLE fmh = mt->FileMappingHandle;
> ```

Seems to be not used (but could be).

---

*[src/windows/win_mmap.c, line 470 \[r3\]](https://reviewable.io:443/reviews/pmem/nvml/1031#-KNfs_hChjGW84sbWW_M:-KNfs_hChjGW84sbWW_N:-647875813) ([raw file](https://github.com/pmem/nvml/blob/589413cb7c39f0161de16ed05f05d79ad746f24c/src/windows/win_mmap.c#L470)):*
> ```C
> 			mte->FileMappingHandle = mt->FileMappingHandle;
> 		else
> 			DuplicateHandle(GetCurrentProcess(), mt->FileMappingHandle,
> ```

Can it fail? Any error handling (or at least assert)?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/nvml/1031#-:-KNfudslzpFAJAKJoWA_:1816364281)*
<!-- Sent from Reviewable.io -->

ChandKV commented:
Thank you @lplewa, added the Ref to the commit description. @krzycz: good question :), if we are building libpmem.dll then and no other project consumes the .c file that calls NtFreeVirtualAddress then yes. If we are building libpmen.lib (static library) then the *binaries* that consume it should link with ntdll.dll. ideally we need it only on the binaries (not static libs). I could see why you ask this, fixed the projects to not link to ntdll where not required, thank you.

Btw, cstyle is complaining about header order in "src/include/libpmemobj++", is there a change to address that one or shall I include those with this, I have the changes handy.

---

Review status: 59 of 93 files reviewed at latest revision, 18 unresolved discussions.

---

*[src/common/set.c, line 1655 \[r1\]](https://reviewable.io:443/reviews/pmem/nvml/1031#-KNab9INnJG1FeL_HYWu:-KNhGtuTyt5qwOCi0V06:257160351) ([raw file](https://github.com/pmem/nvml/blob/6e53c028cd41db4065308f1bd00178402847e68b/src/common/set.c#L1655)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary>
> If the first call to util_map_part() reserves space for all the parts (with VirtualAlloc), then why could subsequent mappings fail? Does it ever happen in practice?
</details>
Because the way we use the reserved space is to unreserve it and then use the same as the hint for the file mapping. In the meantime some other thread could grab that space. NOTE: The other thread need not explicitly ask for this region any request to allocate virtual address could grab this up.

In practice, it depends on how multi threaded the application is w.r.t. reserving virtual address (i.e. allocating memory, mapping files, etc.)

---

*[src/common/set.c, line 1667 \[r1\]](https://reviewable.io:443/reviews/pmem/nvml/1031#-KNa_DQCKpN5G5zn9vfu:-KNhGO7Un2cKKhBKjY3S:1877644203) ([raw file](https://github.com/pmem/nvml/blob/6e53c028cd41db4065308f1bd00178402847e68b/src/common/set.c#L1667)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary>
> `p < p` ???
> I think you should use some other var name for a loop counter.
</details>
Isn't this already fixed in the latest revision, as part of the refactor :)?

---

*[src/common/set.c, line 1673 \[r1\]](https://reviewable.io:443/reviews/pmem/nvml/1031#-KNa_zLZ2MDWqkaobUUn:-KNhHq-50W7j-1j4Zria:-1333304355) ([raw file](https://github.com/pmem/nvml/blob/6e53c028cd41db4065308f1bd00178402847e68b/src/common/set.c#L1673)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary>
> This is confusing. Shouldn't it be: `munmap(addr, rep->repsize);` 
</details>
My bad, stuck with Windows meaning. It should be (rep->repsize - mapsize), i.e. remaining length of the address that we initially reserved. Fixed. Thank you.

---

*[src/common/set.c, line 1705 \[r1\]](https://reviewable.io:443/reviews/pmem/nvml/1031#-KNaaEZ33XWVcvzb6DHn:-KNhIP9aSlHJADKNj12f:1758359711) ([raw file](https://github.com/pmem/nvml/blob/6e53c028cd41db4065308f1bd00178402847e68b/src/common/set.c#L1705)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary>
> as above
</details>
Fixed.

---

*[src/common/set.c, line 1605 \[r3\]](https://reviewable.io:443/reviews/pmem/nvml/1031#-KNfgC9ZbwNSu9egJ2dZ:-KNhFMlz4F-44wDqnElK:-48505393) ([raw file](https://github.com/pmem/nvml/blob/589413cb7c39f0161de16ed05f05d79ad746f24c/src/common/set.c#L1605)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary>
> I'm not sure if retries make any sense on Linux, as it would either succeed on the first attempt (likely) or fail (if there is really no unused region in process address space). In other words, for Linux this counter could be 1, I believe.
> 
> On the other hand, I think we should refactor this code anyway, and reserve the space for _all_ the replicas first, then start to mapping them to memory. It's very unlikely we'll run out of space in case of 64-bit (or 48-bit) addressing, but...
> imagine we have only 3TB contiguous unused region and two replicas - each of 1TB in size. If the first one is placed in a middle of the unused region, the second one could not be mapped...
> 
> Perhaps this is a subject for another patch - just leave an XXX comment on that to not forget about that.
> 
</details>
I thought the retries shouldn't harm if that ever happen in Linux, there by avoiding an #ifndef. Btw, piece of me was thinking that Linux might also have the window. Updated it to have 0 retries (i.e. one attempt) in Linux.

Regarding the replicas, do all the replicas need to be in a single contiguous space? I didn't quite grabbed your comment, but included a note on what I understood.

---

*[src/test/unittest/unittest.ps1, line 74 \[r1\]](https://reviewable.io:443/reviews/pmem/nvml/1031#-KNay4D059ltyRJOtsjS:-KNhEEZFqi1W0JX5ccAi:-2100563572) ([raw file](https://github.com/pmem/nvml/blob/6e53c028cd41db4065308f1bd00178402847e68b/src/test/unittest/unittest.ps1#L74)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary>
> Well, remember that util_parse_size() is a different creature - it accepts size units as described in our man pages. This function however, may pass the size (unmodified) to `truncate` command (on Linux), which has slightly different semantics than util_parse_size().
> Anyway, the new version looks good, and it seems to be more or less compatible with `truncate`.
</details>
Thank you @lplewa. Let me leave convert_to_bytes() to be case insensitive for now. And to avoid human err, I'm thinking that convert_to_bytes() should fail if it's passed with bytes greater than 1024. The reason being that, if the caller or the caller's caller is not so careful in preserving the string say "1mb" and instead let powershell expand it to "1048576". We would then notify the caller giving the developer a chance to fix the same. Let me know how you both like it.

Also, I fixed a bug in the previous version.

---

*[src/test/win_mmap/win_mmap.c, line 265 \[r3\]](https://reviewable.io:443/reviews/pmem/nvml/1031#-KNfeY23lD12gM39q6mJ:-KNhN34yRUf694PlYsRp:-2040547124) ([raw file](https://github.com/pmem/nvml/blob/589413cb7c39f0161de16ed05f05d79ad746f24c/src/test/win_mmap/win_mmap.c#L265)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary>
> Why `0`? And why not `check_mapping()` ?
</details>
My bad stuck with Windows mind set, where 0 is rest of the mapping. Let me fix this.

For this test, the intention is to capture that we are failing it in Windows implementation for now. I can try to see what mapping does Linux creates if we map beyond file size, just that I thought it's less important. Let me check that once :).

---

*[src/windows/win_mmap.c, line 153 \[r3\]](https://reviewable.io:443/reviews/pmem/nvml/1031#-KNfu9ycOqGx72IBB65B:-KNhPlOMqq4GgSlSaQFl:-499146009) ([raw file](https://github.com/pmem/nvml/blob/589413cb7c39f0161de16ed05f05d79ad746f24c/src/windows/win_mmap.c#L153)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary>
> ... -- (internal) ...
</details>
Done.

---

*[src/windows/win_mmap.c, line 155 \[r3\]](https://reviewable.io:443/reviews/pmem/nvml/1031#-KNfu8D10MEwiUb3kBM0:-KNhPliCqMxc8JlnnaFD:-499146009) ([raw file](https://github.com/pmem/nvml/blob/589413cb7c39f0161de16ed05f05d79ad746f24c/src/windows/win_mmap.c#L155)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary>
> static int
</details>
Done.

---

*[src/windows/win_mmap.c, line 284 \[r3\]](https://reviewable.io:443/reviews/pmem/nvml/1031#-KNfkkFP-Ect-6FT8e7Q:-KNhPp4V35YKibEMsxWo:-1753633887) ([raw file](https://github.com/pmem/nvml/blob/589413cb7c39f0161de16ed05f05d79ad746f24c/src/windows/win_mmap.c#L284)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary>
> MAP_ANON
</details>
Done :), thank you.

---

*[src/windows/win_mmap.c, line 344 \[r3\]](https://reviewable.io:443/reviews/pmem/nvml/1031#-KNfphNK7DhiYlgbnN5l:-KNhQB2KzNws4ki9HNtB:-1854434631) ([raw file](https://github.com/pmem/nvml/blob/589413cb7c39f0161de16ed05f05d79ad746f24c/src/windows/win_mmap.c#L344)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary>
> Not needed, I think, as addr == reserved_addr at this point.
</details>
Not when addr is NULL, right?

---

*[src/windows/win_mmap.c, line 401 \[r3\]](https://reviewable.io:443/reviews/pmem/nvml/1031#-KNfnhMSQvZ0GygzAMpR:-KNhQjazzJWMHK0i2We9:-546001267) ([raw file](https://github.com/pmem/nvml/blob/589413cb7c39f0161de16ed05f05d79ad746f24c/src/windows/win_mmap.c#L401)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary>
> Isn't an assert enough? Seems like it's already handled above (except for checking if `base == addr`).
</details>
Not sure if I see what you are saying. The above checks handle the case where MapViewOfFileEx can't map and returns NULL, i.e. they handle the case where "base == NULL". We are checking for the case where MapViewOfFileEx can allocate but not at the given hint. Am I missing something here?

---

*[src/windows/win_mmap.c, line 444 \[r3\]](https://reviewable.io:443/reviews/pmem/nvml/1031#-KNfsMVCvx2bfoCqPTEB:-KNhRT_6AT0hKAc9r3PL:2117696956) ([raw file](https://github.com/pmem/nvml/blob/589413cb7c39f0161de16ed05f05d79ad746f24c/src/windows/win_mmap.c#L444)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary>
> Seems to be not used (but could be).
</details>
Put that to good use ;). I see more bugs there, so opened a trello card.

---

*[src/windows/win_mmap.c, line 470 \[r3\]](https://reviewable.io:443/reviews/pmem/nvml/1031#-KNfs_hChjGW84sbWW_M:-KNhUUpfUPIB88rD-076:1575642829) ([raw file](https://github.com/pmem/nvml/blob/589413cb7c39f0161de16ed05f05d79ad746f24c/src/windows/win_mmap.c#L470)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary>
> Can it fail? Any error handling (or at least assert)?
</details>
Thanks for catching this, it can fail. Fixed.

---

*[src/test/blk_rw/err0w.log.match, line  \[r1\]](https://reviewable.io:443/reviews/pmem/nvml/1031#-KNaWV-8N4B56jU2HdaU:-KNhIfCs4WRGmDK-IRTq:-1537651880) ([raw file](https://github.com/pmem/nvml/blob/a1b4e45a97ba5f200302af19bf137d53cda5b88d/src/test/blk_rw/err0w.log.match#L)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary>
> I mean, for tests 1-7 output is the same on Linux and Windows, so we don't need to have two match files for each of those (with and without "w" suffix). This is required only for tests 8-9. 
</details>
Sorry about that, it missed my sight when unifying the tests that are initially made different. Removed them.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/nvml/1031#-:-KNh_YKy3iCJLhy-zAt0:241951915)*
<!-- Sent from Reviewable.io -->

krzycz commented:
Cstyle should not check C++ files, as they are using different coding style (clang-format). CSTYLE.ps1 should scan only *.c and *.h files, so I have no idea why is it touching *hpp files.

---

Reviewed 31 of 42 files at r4, 20 of 20 files at r7.
Review status: all files reviewed at latest revision, 9 unresolved discussions, some commit checks failed.

---

*[src/common/set.c, line 1655 \[r1\]](https://reviewable.io:443/reviews/pmem/nvml/1031#-KNab9INnJG1FeL_HYWu:-KNkdKatRqJpzYiv8M6D:-584367351) ([raw file](https://github.com/pmem/nvml/blob/6e53c028cd41db4065308f1bd00178402847e68b/src/common/set.c#L1655)):*
<details><summary><i>Previously, ChandKV (Chandra Kumar) wrote…</i></summary>
> Because the way we use the reserved space is to unreserve it and then use the same as the hint for the file mapping. In the meantime some other thread could grab that space. NOTE: The other thread need not explicitly ask for this region any request to allocate virtual address could grab this up.
> 
> In practice, it depends on how multi threaded the application is w.r.t. reserving virtual address (i.e. allocating memory, mapping files, etc.)
</details>
Well, right! I didn't think about MT scenario. 

---

*[src/common/set.c, line 1667 \[r1\]](https://reviewable.io:443/reviews/pmem/nvml/1031#-KNa_DQCKpN5G5zn9vfu:-KNkeEye-2Hw_l9bMIhG:-1138413255) ([raw file](https://github.com/pmem/nvml/blob/6e53c028cd41db4065308f1bd00178402847e68b/src/common/set.c#L1667)):*
<details><summary><i>Previously, ChandKV (Chandra Kumar) wrote…</i></summary>
> Isn't this already fixed in the latest revision, as part of the refactor :)?
</details>
Yes, it is. But I put my comment before that.

---

*[src/common/set.c, line 1605 \[r3\]](https://reviewable.io:443/reviews/pmem/nvml/1031#-KNfgC9ZbwNSu9egJ2dZ:-KNparxIQSNKeIOFW4rY:-1421993022) ([raw file](https://github.com/pmem/nvml/blob/589413cb7c39f0161de16ed05f05d79ad746f24c/src/common/set.c#L1605)):*
<details><summary><i>Previously, ChandKV (Chandra Kumar) wrote…</i></summary>
> I thought the retries shouldn't harm if that ever happen in Linux, there by avoiding an #ifndef. Btw, piece of me was thinking that Linux might also have the window. Updated it to have 0 retries (i.e. one attempt) in Linux.
> 
> Regarding the replicas, do all the replicas need to be in a single contiguous space? I didn't quite grabbed your comment, but included a note on what I understood.
> 
</details>
No, the replicas do not need to be in a single contiguous space. It's just about the optimal utilization of the address space (reduced fragmentation). Especially when we're running out of space. 

---

*[src/test/unittest/unittest.ps1, line 74 \[r1\]](https://reviewable.io:443/reviews/pmem/nvml/1031#-KNay4D059ltyRJOtsjS:-KNpdQ-WBzZS-3hsMVqz:-1675198818) ([raw file](https://github.com/pmem/nvml/blob/6e53c028cd41db4065308f1bd00178402847e68b/src/test/unittest/unittest.ps1#L74)):*
<details><summary><i>Previously, ChandKV (Chandra Kumar) wrote…</i></summary>
> Thank you @lplewa. Let me leave convert_to_bytes() to be case insensitive for now. And to avoid human err, I'm thinking that convert_to_bytes() should fail if it's passed with bytes greater than 1024. The reason being that, if the caller or the caller's caller is not so careful in preserving the string say "1mb" and instead let powershell expand it to "1048576". We would then notify the caller giving the developer a chance to fix the same. Let me know how you both like it.
> 
> Also, I fixed a bug in the previous version.
</details>
I was going to protest against not allowing to specify size larger than 1K in bytes, as this could be useful for some tests (when you want to check the library behavior in case of unaligned file size). But, since there is support for "b" suffix, I'm fine with that.

---

*[src/test/unittest/unittest.ps1, line 80 \[r2\]](https://reviewable.io:443/reviews/pmem/nvml/1031#-KNpeHjh7KNEIktnLFj9:-KNpeHjlnlsBOCPXZvvn:-1180691948) ([raw file](https://github.com/pmem/nvml/blob/4ff2094c960ded839e7306bd67ccef64b3f31149/src/test/unittest/unittest.ps1#L80)):*
> ```PowerShell
>     if ($size.ToLower().EndsWith("b")) {
>         $size = [int64]($size.Substring(0, $size.Length - 1))
>     } elseif ($size.ToLower().EndsWith("k")) {
> ```

;-) Yeah, I forgot to put a comment on that, but also had a feeling that longest suffixes should be checked first...

---

*[src/windows/win_mmap.c, line 344 \[r3\]](https://reviewable.io:443/reviews/pmem/nvml/1031#-KNfphNK7DhiYlgbnN5l:-KNkflneADXziSQZoOdv:-865860947) ([raw file](https://github.com/pmem/nvml/blob/589413cb7c39f0161de16ed05f05d79ad746f24c/src/windows/win_mmap.c#L344)):*
<details><summary><i>Previously, ChandKV (Chandra Kumar) wrote…</i></summary>
> Not when addr is NULL, right?
</details>
Right. :-)

---

*[src/windows/win_mmap.c, line 401 \[r3\]](https://reviewable.io:443/reviews/pmem/nvml/1031#-KNfnhMSQvZ0GygzAMpR:-KNkfwrbheAtG_8dEtO6:-467614568) ([raw file](https://github.com/pmem/nvml/blob/589413cb7c39f0161de16ed05f05d79ad746f24c/src/windows/win_mmap.c#L401)):*
<details><summary><i>Previously, ChandKV (Chandra Kumar) wrote…</i></summary>
> Not sure if I see what you are saying. The above checks handle the case where MapViewOfFileEx can't map and returns NULL, i.e. they handle the case where "base == NULL". We are checking for the case where MapViewOfFileEx can allocate but not at the given hint. Am I missing something here?
</details>
I see, but is it possible (assuming `addr != NULL`) that MapViewOfFileEx() returns an address that is different than hint address?
Correct me if I'm wrong, but based on my experiments with MapViewOfFileEx(), it looks like providing a hint address to this function works more or less as MAP_FIXED flag on Linux. The difference is that if the region pointed by hint address is already used, mmap() + MAP_FIXED would override it with a new mapping (on Linux), while MapViewOffileEx() would fail and return NULL. IOW, if it succeeds, then we can assume that `base == addr`.

---

*[utils/check_license/check-license.c, line 349 \[r7\]](https://reviewable.io:443/reviews/pmem/nvml/1031#-KNpfPrjL1JmswzI0eLz:-KNpfPrjL1JmswzI0eM-:608588156) ([raw file](https://github.com/pmem/nvml/blob/b71602b36c4302c370542651707956db83ef0307/utils/check_license/check-license.c#L349)):*
> ```C
> 		copyright += COPYRIGHT_LEN;
> 
> 		/* skip the copyright symbol '(c)' if any */
> ```

I was going to suggest that we should also check for "Intel" in copyright string, and check the format only in such case. Other contributors (including MSFT) may use their own copyright format, I think.
But since you've already made the changes...

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/nvml/1031#-:-KNpgQgtSYCMuGs6cwjA:-290255184)*
<!-- Sent from Reviewable.io -->

ChandKV commented:
Just so it is clear, CSTYLE.PS1 is not complaining about them, I see only when running "make cstyle" in Linux.

---

Review status: 69 of 94 files reviewed at latest revision, 10 unresolved discussions.

---

*[src/common/set.c, line 1667 \[r1\]](https://reviewable.io:443/reviews/pmem/nvml/1031#-KNa_DQCKpN5G5zn9vfu:-KNrBr6Aioptbk-FUu7K:-2144865024) ([raw file](https://github.com/pmem/nvml/blob/6e53c028cd41db4065308f1bd00178402847e68b/src/common/set.c#L1667)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary>
> Yes, it is. But I put my comment before that.
</details>
Yeah, I saw just wanted to say it's fixed.

---

*[src/common/set.c, line 1605 \[r3\]](https://reviewable.io:443/reviews/pmem/nvml/1031#-KNfgC9ZbwNSu9egJ2dZ:-KNrAQZSoeneDMKgRvno:-1145475279) ([raw file](https://github.com/pmem/nvml/blob/589413cb7c39f0161de16ed05f05d79ad746f24c/src/common/set.c#L1605)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary>
> No, the replicas do not need to be in a single contiguous space. It's just about the optimal utilization of the address space (reduced fragmentation). Especially when we're running out of space. 
</details>
I see. I'm not very keen yet, but updated the note appropriately.

---

*[src/test/unittest/unittest.ps1, line 74 \[r1\]](https://reviewable.io:443/reviews/pmem/nvml/1031#-KNay4D059ltyRJOtsjS:-KNrC-MDKcECl9byiHYn:-499205157) ([raw file](https://github.com/pmem/nvml/blob/6e53c028cd41db4065308f1bd00178402847e68b/src/test/unittest/unittest.ps1#L74)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary>
> I was going to protest against not allowing to specify size larger than 1K in bytes, as this could be useful for some tests (when you want to check the library behavior in case of unaligned file size). But, since there is support for "b" suffix, I'm fine with that.
</details>
Agree with you, there needs to be an override else there is no point for such a limitation. As you noted "b" suffix is for the same.

---

*[src/test/unittest/unittest.ps1, line 80 \[r2\]](https://reviewable.io:443/reviews/pmem/nvml/1031#-KNpeHjh7KNEIktnLFj9:-KNrCBpUWoaLS_JlyxMo:2001608915) ([raw file](https://github.com/pmem/nvml/blob/4ff2094c960ded839e7306bd67ccef64b3f31149/src/test/unittest/unittest.ps1#L80)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary>
> ;-) Yeah, I forgot to put a comment on that, but also had a feeling that longest suffixes should be checked first...
</details>
:) yes...

---

*[src/windows/win_mmap.c, line 401 \[r3\]](https://reviewable.io:443/reviews/pmem/nvml/1031#-KNfnhMSQvZ0GygzAMpR:-KNrH_CGs20V1jOmLp-I:950726987) ([raw file](https://github.com/pmem/nvml/blob/589413cb7c39f0161de16ed05f05d79ad746f24c/src/windows/win_mmap.c#L401)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary>
> I see, but is it possible (assuming `addr != NULL`) that MapViewOfFileEx() returns an address that is different than hint address?
> Correct me if I'm wrong, but based on my experiments with MapViewOfFileEx(), it looks like providing a hint address to this function works more or less as MAP_FIXED flag on Linux. The difference is that if the region pointed by hint address is already used, mmap() + MAP_FIXED would override it with a new mapping (on Linux), while MapViewOffileEx() would fail and return NULL. IOW, if it succeeds, then we can assume that `base == addr`.
</details>
I see you now. You are correct, it shouldn't. The similar check in util_map_part() made me add this. Removed the same. Thank you.

---

*[utils/check_license/check-license.c, line 349 \[r7\]](https://reviewable.io:443/reviews/pmem/nvml/1031#-KNpfPrjL1JmswzI0eLz:-KNrI3cioW8DiK4LYFU2:1494068475) ([raw file](https://github.com/pmem/nvml/blob/b71602b36c4302c370542651707956db83ef0307/utils/check_license/check-license.c#L349)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary>
> I was going to suggest that we should also check for "Intel" in copyright string, and check the format only in such case. Other contributors (including MSFT) may use their own copyright format, I think.
> But since you've already made the changes...
</details>
This is very interesting :). I did precisely what you thought of saying. But then in yesterday's call @lplewa suggested that Microsoft also use the year as BSD license requires that and remove the (c) string is possible. So I had a back and forth conversation with my LCA contact to get a best fit :). I think we can go with this for now.

Btw, I assume it's OK relax the (c) check for Intel, if that's not true we might need to check based on the organisation.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/nvml/1031#-:-KNrAAaIdwNvQfFkFsk0:-1056871620)*
<!-- Sent from Reviewable.io -->

krzycz commented:
Reviewed 25 of 25 files at r8.
Review status: all files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---

*[utils/check_license/check-license.c, line 349 \[r7\]](https://reviewable.io:443/reviews/pmem/nvml/1031#-KNpfPrjL1JmswzI0eLz:-KNrug21564Xm18low88:-703859386) ([raw file](https://github.com/pmem/nvml/blob/b71602b36c4302c370542651707956db83ef0307/utils/check_license/check-license.c#L349)):*
<details><summary><i>Previously, ChandKV (Chandra Kumar) wrote…</i></summary>
> This is very interesting :). I did precisely what you thought of saying. But then in yesterday's call @lplewa suggested that Microsoft also use the year as BSD license requires that and remove the (c) string is possible. So I had a back and forth conversation with my LCA contact to get a best fit :). I think we can go with this for now.
> 
> Btw, I assume it's OK relax the (c) check for Intel, if that's not true we might need to check based on the organisation.
</details>
Well, we were told by our legal that "(c)" is not necessary, and we have removed it one day from all the files. On the other hand, it's just a string - for me, the most important part are the words "copyright" and the copyright holder/company name, and the license of course.  "(c)", years, etc. are just details.
But... that's only my opinion - I'm not a lawyer - so, we try to stick to the instructions we got from our legal. ;-)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/nvml/1031)*
<!-- Sent from Reviewable.io -->

krzycz commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: all files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/nvml/1031#-:-KNrvvrDaa9FKD7i3R77:1417088721)*
<!-- Sent from Reviewable.io -->

krzycz commented:
Merged by hand (resolved conflicts and fixed some issues with srcversion.h).
